bin/varnishd/cache/cache_ban.c

```diff
@@ -769,9 +769,77 @@ ccf_ban_list(struct cli *cli, const char * const *av, void *priv)
 	Lck_Unlock(&ban_mtx);
 }
 
+static void v_matchproto_(cli_func_t)
+ccf_ban_list_json(struct cli *cli, const char * const *av, void *priv)
+{
+	struct ban *b, *bl;
+	int64_t o;
+	int n = 0;
+	int ocs;
+
+	(void)priv;
+
+	Lck_Lock(&ban_mtx);
+	bl = VTAILQ_LAST(&ban_head, banhead_s);
+	bl->refcount++;
+	Lck_Unlock(&ban_mtx);
+
+	VCLI_JSON_begin(cli, 2, av);
+	VCLI_Out(cli, ",\n");
+	VTAILQ_FOREACH(b, &ban_head, list) {
+		o = bl == b ? 1 : 0;
+		VCLI_Out(cli, "%s", n ? ",\n" : "");
+		n++;
+		VCLI_Out(cli, "{\n");
+		VSB_indent(cli->sb, 2);
+		VCLI_Out(cli, "\"time\": %.6f,\n", ban_time(b->spec));
+		VCLI_Out(cli, "\"refs\": %ju,\n", (intmax_t)(b->refcount - o));
+		VCLI_Out(cli, "\"completed\": %s,\n",
+			 b->flags & BANS_FLAG_COMPLETED ? "true" : "false");
+		VCLI_Out(cli, "\"spec\": \"");
+		ban_render(cli, b->spec);
```

Comment:
@nigoroll none of the test outputs required it. The only thing I see in ban_render() that might require quoting is the output of the ban argument: ``VCLI_Out(cli, "%s", bt.arg2);`` (currently line 719 in cache_ban.c). Could that potentially contain something that needs to be quoted? (You know bans better than I do.) If you can write up an example it would help a lot, then I'd add it to the tests and write up the code.

Generally I'm trying to avoid rewriting existing code like ban_render() where possible. But if the quoting can be limited to that one line, it won't be too obtrusive

Reply:
Ok just tried a test using a ban expression in which the arg is quoted (and will be adding it to c00019.vtc):
```
# Ban expression with quoting
varnish v1 -cliok {ban req.url ~ "BAR"}
varnish v1 -cliok "ban.list"
varnish v1 -clijson "ban.list -j"
```

The ban expression with quoting is permitted (note: on my reading, vcl(7) implies that it would have been an error, but it evidently isn't), but the quotation marks are removed in the ``ban.list`` output.

From this I infer that no part of ``ban.list`` output requires quoting in the JSON rendering.

@nigoroll I'm still convinced that you know bans better than I do, so I don't mind if you can prove me wrong.